integration/tests/basic_functionality/basic_deployment.rb

```diff
@@ -0,0 +1,144 @@
+  on(master, "#{r10k_fqp} deploy environment production --verbose=info") do |result|
+    assert(result.stdout =~ /.*Deploying module to .*apache.*/, 'Did not log apache deployment')
+    assert(result.stdout =~ /.*Deploying module to .*stdlib.*/, 'Did not log stdlib deployment')
+  end
+  on(master, "test -f #{test_path}", accept_all_exit_codes: true) do |result|
+    assert(result.exit_code == 0, 'Expected module in Puppetfile was not installed')
+  end
+
+  and_stdlib_is_correct
+end
+
+original_apache_info = JSON.parse(on(master, "cat #{test_path}").stdout)
+
+step 'Test second run of deploy updates control repo, but not modules' do
+  puppetfile2 =<<-EOS
+    mod 'puppetlabs/apache', :latest
+    mod 'puppetlabs/stdlib', '8.0.0'
+    EOS
+
+  git_on(master, 'checkout production', control_repo_worktree)
+  create_remote_file(master, "#{control_repo_worktree}/Puppetfile", puppetfile2)
+  git_add_commit_push(master, 'production', 'add Puppetfile for Basic Deployment test', control_repo_worktree)
+
+  on(master, "#{r10k_fqp} deploy environment production --verbose=info") do |result|
+    refute(result.stdout =~ /.*Deploying module to .*apache.*/, 'Inappropriately updated apache')
+    refute(result.stdout =~ /.*Deploying module to .*stdlib.*/, 'Inappropriately updated stdlib')
+  end
+  on(master, "test -f #{test_path}", accept_all_exit_codes: true) do |result|
+    assert(result.exit_code == 0, 'Expected module content in Puppetfile was inappropriately purged')
+  end
+
+  new_apache_info = JSON.parse(on(master, "cat #{test_path}").stdout)
+  on(master, "cat #{env_path}/production/Puppetfile | grep 5.0.0", accept_all_exit_codes: true) do |result|
+    assert(result.exit_code == 0, 'Puppetfile not updated on subsequent r10k deploys')
+  end
+
+  assert(original_apache_info['version'] == new_apache_info['version'] &&
+         new_apache_info['version'] == '0.10.0',
+         'Module content updated on subsequent r10k invocations w/o providing --modules')
+
+  and_stdlib_is_correct
+end
+
+step 'Test --modules updates modules' do
+  on(master, "#{r10k_fqp} deploy environment production --modules --verbose=info") do |result|
+    assert(result.stdout =~ /.*Deploying module to .*apache.*/, 'Did not log apache deployment')
+    assert(result.stdout =~ /.*Deploying module to .*stdlib.*/, 'Did not log stdlib deployment')
+  end
+  on(master, "test -f #{test_path}", accept_all_exit_codes: true) do |result|
+    assert(result.exit_code == 0, 'Expected module content in Puppetfile was inappropriately purged')
+  end
+
+  new_apache_info = JSON.parse(on(master, "cat #{test_path}").stdout)
+  apache_major_version = new_apache_info['version'].split('.').first.to_i
+  assert(apache_major_version > 5, 'Module not updated correctly using --modules')
+
+  and_stdlib_is_correct
+end
+
+step 'Test --modules --incremental updates changed modules' do
+  on(master, "#{r10k_fqp} deploy environment production --modules --incremental --verbose=debug1") do |result|
```

Comment:
Without a new Puppetfile, this deploy isn't actually changing anything, is it? Is there a better way we can test this behavior other than checking the log messages? Just trying to think in terms of actual workflows, since I don't imagine someone would run a deploy with `--modules` and then again with `--modules --incremental` without at least making some sort of change in between?

Reply:
I'll add a module that is updated in this invocation to test the changeable vs static, changed vs static, unchanged. Let me know if you'd like more permutations of static version checking (fwiw, there is some unit tests around that: https://github.com/puppetlabs/r10k/commit/6f46384633531d82df10d490183259eaccd8ad93#diff-1bc365ddca4e6e4a7ba95ff4d655c036faa16ec1025e35bc559476265a5719aaR344-R389)

Reply:
I think what you added looks good! I'm satisfied with the various permutations we test here.